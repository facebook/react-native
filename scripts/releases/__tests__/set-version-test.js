/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

const {setVersion} = require('../set-version');
const path = require('path');

jest.mock('../../consts', () => ({
  REPO_ROOT: path.join(__dirname, '__fixtures__', 'set-version'),
  PACKAGES_DIR: path.join(__dirname, '__fixtures__', 'set-version', 'packages'),
  REACT_NATIVE_PACKAGE_DIR: path.join(
    __dirname,
    '__fixtures__',
    'set-version',
    'packages',
    'react-native',
  ),
}));

let customWriteFileExpect = null;
const writeFileMock = jest.fn().mockImplementation((filePath, content) => {
  if (customWriteFileExpect != null) {
    customWriteFileExpect(filePath, content);
  }

  const normalizedFilePath = path
    .relative(path.join(__dirname, '__fixtures__'), filePath)
    .split(path.sep)
    .join('/');

  if (!filePath.endsWith('package.json')) {
    // Updated source and build files are already validated in the tests for
    // `set-rn-artifacts-version.js`. We also want to avoid polluting this test's
    // snapshots with \@\generated.
    expect('[omitted]').toMatchSnapshot(normalizedFilePath);
    return;
  }

  expect(content).toMatchSnapshot(normalizedFilePath);
});

describe('setVersion', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  beforeAll(() => {
    jest.mock('fs', () => {
      // $FlowIgnore[underconstrained-implicit-instantiation]
      const originalFs = jest.requireActual('fs');

      return {
        ...originalFs,
        writeFileSync: writeFileMock,
        promises: {
          ...originalFs.promises,
          writeFile: writeFileMock,
        },
      };
    });
  });

  test('updates monorepo for release-candidate', async () => {
    await setVersion('0.80.0-rc.3');
  });

  test('updates monorepo for stable version', async () => {
    await setVersion('0.80.1');
  });

  test('updates monorepo for nightly', async () => {
    await setVersion('0.81.0-nightly-29282302-abcd1234');
  });

  test('updates monorepo on main after release cut', async () => {
    customWriteFileExpect = (filePath /*: string */, content /*: string */) => {
      const reactNativePath = path.join('react-native', 'package.json');
      if (filePath.endsWith(reactNativePath)) {
        expect(JSON.parse(content).version).toBe('1000.0.0');
      }
      const templatePath = path.join(
        'react-native',
        'template',
        'package.json',
      );
      if (filePath.endsWith(templatePath)) {
        expect(JSON.parse(content).dependencies['react-native']).toBe(
          '1000.0.0',
        );
      }
    };

    await setVersion('0.82.0-main', true);
  });
});
