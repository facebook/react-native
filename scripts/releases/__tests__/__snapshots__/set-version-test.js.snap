// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setVersion updates monorepo for nightly: set-version/package.json 1`] = `
"{
  \\"name\\": \\"@react-native/monorepo\\",
  \\"private\\": true,
  \\"version\\": \\"1000.0.0\\",
  \\"devDependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.81.0-nightly-29282302-abcd1234\\"
  }
}
"
`;

exports[`setVersion updates monorepo for nightly: set-version/packages/monorepo-pkg-a/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-a\\",
  \\"version\\": \\"0.81.0-nightly-29282302-abcd1234\\",
  \\"description\\": \\"@monorepo/pkg-a\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-b\\": \\"0.81.0-nightly-29282302-abcd1234\\",
    \\"@monorepo/other\\": \\"0.0.1\\"
  },
  \\"devDependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.81.0-nightly-29282302-abcd1234\\"
  }
}
"
`;

exports[`setVersion updates monorepo for nightly: set-version/packages/monorepo-pkg-b/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-b\\",
  \\"version\\": \\"0.81.0-nightly-29282302-abcd1234\\",
  \\"description\\": \\"@monorepo/pkg-b\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.81.0-nightly-29282302-abcd1234\\",
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo for nightly: set-version/packages/monorepo-pkg-c/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-c\\",
  \\"version\\": \\"0.81.0-nightly-29282302-abcd1234\\",
  \\"description\\": \\"@monorepo/pkg-c\\",
  \\"dependencies\\": {
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo for nightly: set-version/packages/react-native/Libraries/Core/ReactNativeVersion.js 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for nightly: set-version/packages/react-native/React/Base/RCTVersion.m 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for nightly: set-version/packages/react-native/ReactAndroid/gradle.properties 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for nightly: set-version/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/ReactNativeVersion.java 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for nightly: set-version/packages/react-native/ReactCommon/cxxreact/ReactNativeVersion.h 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for nightly: set-version/packages/react-native/package.json 1`] = `
"{
  \\"name\\": \\"react-native\\",
  \\"version\\": \\"0.81.0-nightly-29282302-abcd1234\\",
  \\"description\\": \\"fake react native package\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-a\\": \\"0.81.0-nightly-29282302-abcd1234\\",
    \\"@monorepo/pkg-b\\": \\"0.81.0-nightly-29282302-abcd1234\\",
    \\"@monorepo/pkg-c\\": \\"0.81.0-nightly-29282302-abcd1234\\",
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo for release-candidate: set-version/package.json 1`] = `
"{
  \\"name\\": \\"@react-native/monorepo\\",
  \\"private\\": true,
  \\"version\\": \\"1000.0.0\\",
  \\"devDependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.80.0-rc.3\\"
  }
}
"
`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/monorepo-pkg-a/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-a\\",
  \\"version\\": \\"0.80.0-rc.3\\",
  \\"description\\": \\"@monorepo/pkg-a\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-b\\": \\"0.80.0-rc.3\\",
    \\"@monorepo/other\\": \\"0.0.1\\"
  },
  \\"devDependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.80.0-rc.3\\"
  }
}
"
`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/monorepo-pkg-b/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-b\\",
  \\"version\\": \\"0.80.0-rc.3\\",
  \\"description\\": \\"@monorepo/pkg-b\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.80.0-rc.3\\",
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/monorepo-pkg-c/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-c\\",
  \\"version\\": \\"0.80.0-rc.3\\",
  \\"description\\": \\"@monorepo/pkg-c\\",
  \\"dependencies\\": {
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/react-native/Libraries/Core/ReactNativeVersion.js 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/react-native/React/Base/RCTVersion.m 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/react-native/ReactAndroid/gradle.properties 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/ReactNativeVersion.java 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/react-native/ReactCommon/cxxreact/ReactNativeVersion.h 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for release-candidate: set-version/packages/react-native/package.json 1`] = `
"{
  \\"name\\": \\"react-native\\",
  \\"version\\": \\"0.80.0-rc.3\\",
  \\"description\\": \\"fake react native package\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-a\\": \\"0.80.0-rc.3\\",
    \\"@monorepo/pkg-b\\": \\"0.80.0-rc.3\\",
    \\"@monorepo/pkg-c\\": \\"0.80.0-rc.3\\",
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo for stable version: set-version/package.json 1`] = `
"{
  \\"name\\": \\"@react-native/monorepo\\",
  \\"private\\": true,
  \\"version\\": \\"1000.0.0\\",
  \\"devDependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.80.1\\"
  }
}
"
`;

exports[`setVersion updates monorepo for stable version: set-version/packages/monorepo-pkg-a/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-a\\",
  \\"version\\": \\"0.80.1\\",
  \\"description\\": \\"@monorepo/pkg-a\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-b\\": \\"0.80.1\\",
    \\"@monorepo/other\\": \\"0.0.1\\"
  },
  \\"devDependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.80.1\\"
  }
}
"
`;

exports[`setVersion updates monorepo for stable version: set-version/packages/monorepo-pkg-b/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-b\\",
  \\"version\\": \\"0.80.1\\",
  \\"description\\": \\"@monorepo/pkg-b\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.80.1\\",
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo for stable version: set-version/packages/monorepo-pkg-c/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-c\\",
  \\"version\\": \\"0.80.1\\",
  \\"description\\": \\"@monorepo/pkg-c\\",
  \\"dependencies\\": {
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo for stable version: set-version/packages/react-native/Libraries/Core/ReactNativeVersion.js 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for stable version: set-version/packages/react-native/React/Base/RCTVersion.m 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for stable version: set-version/packages/react-native/ReactAndroid/gradle.properties 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for stable version: set-version/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/ReactNativeVersion.java 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for stable version: set-version/packages/react-native/ReactCommon/cxxreact/ReactNativeVersion.h 1`] = `"[omitted]"`;

exports[`setVersion updates monorepo for stable version: set-version/packages/react-native/package.json 1`] = `
"{
  \\"name\\": \\"react-native\\",
  \\"version\\": \\"0.80.1\\",
  \\"description\\": \\"fake react native package\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-a\\": \\"0.80.1\\",
    \\"@monorepo/pkg-b\\": \\"0.80.1\\",
    \\"@monorepo/pkg-c\\": \\"0.80.1\\",
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo on main after release cut: set-version/package.json 1`] = `
"{
  \\"name\\": \\"@react-native/monorepo\\",
  \\"private\\": true,
  \\"version\\": \\"1000.0.0\\",
  \\"devDependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.82.0-main\\"
  }
}
"
`;

exports[`setVersion updates monorepo on main after release cut: set-version/packages/monorepo-pkg-a/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-a\\",
  \\"version\\": \\"0.82.0-main\\",
  \\"description\\": \\"@monorepo/pkg-a\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-b\\": \\"0.82.0-main\\",
    \\"@monorepo/other\\": \\"0.0.1\\"
  },
  \\"devDependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.82.0-main\\"
  }
}
"
`;

exports[`setVersion updates monorepo on main after release cut: set-version/packages/monorepo-pkg-b/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-b\\",
  \\"version\\": \\"0.82.0-main\\",
  \\"description\\": \\"@monorepo/pkg-b\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-c\\": \\"0.82.0-main\\",
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo on main after release cut: set-version/packages/monorepo-pkg-c/package.json 1`] = `
"{
  \\"name\\": \\"@monorepo/pkg-c\\",
  \\"version\\": \\"0.82.0-main\\",
  \\"description\\": \\"@monorepo/pkg-c\\",
  \\"dependencies\\": {
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;

exports[`setVersion updates monorepo on main after release cut: set-version/packages/react-native/package.json 1`] = `
"{
  \\"name\\": \\"react-native\\",
  \\"version\\": \\"1000.0.0\\",
  \\"description\\": \\"fake react native package\\",
  \\"dependencies\\": {
    \\"@monorepo/pkg-a\\": \\"0.82.0-main\\",
    \\"@monorepo/pkg-b\\": \\"0.82.0-main\\",
    \\"@monorepo/pkg-c\\": \\"0.82.0-main\\",
    \\"metro-config\\": \\"^0.80.10\\",
    \\"metro-runtime\\": \\"^0.80.10\\"
  }
}
"
`;
